Aggregation helpers

benallfree opened the pull request:
I tried a few different strategies. First I thought this would be a knex monkey patch, but no. It's a `QueryBuilder` update. I think the best strategy is to wrap whatever the knex query is rather than attempting to modify it by adding a `.count()` to it. 

Example:

```javascript
await User.rowsCount()
```

Produces

```sql
select count(*) as "total" from (select * from "users") as "__count"
```
That way, no matter how complex the original query is, it just gets turned into a subquery and a row count is returned.

Since this terminates `QueryBuilder` chaining without actually calling `fetch()`, I don't know what the implications are. Should I duplicate what the `fetch()` method does?

benallfree commented:
Related docs https://github.com/adonisjs/docs/pull/175

benallfree commented:
Ready to merge along with https://github.com/adonisjs/docs/pull/175

thetutlage commented:
What you think of moving actual query builder methods from `src/Lucid/QueryBuilder/index.js` to `src/Database/index.js`.

This way it doesn't matter whether I am using Lucid or the actual query builder from Database provider.

benallfree commented:
I like it. And combine with docs here? http://dev.adonisjs.com/docs/4.0/query-builder#_aggregates

thetutlage commented:
Yeah, move the docs there

benallfree commented:
Ok that was a good idea. I still need to update docs, but please take a look at this first.

One question I have... I don't understand why `QueryBuilder` needs to explicitly wrap these functions. https://github.com/adonisjs/adonis-lucid/pull/205/files#diff-13ef6d505395ea241aa8a59803cc7315R1347

`await User.query().getCount()` does call `KnexQueryBuilder.getCount()` but it returns the result from `QueryBuilder.query` (which is `select * from users`) instead of the result from `getCount()`.

benallfree commented:
Hold on this..bad push

benallfree commented:
Ok this is working great now.

@thetutlage Please carefully examine my change to https://github.com/adonisjs/adonis-lucid/pull/205/files#diff-254364470b0e90c9cfcca5ff2433e8f2. I made this change so I wouldn't have to add `getXXX()` wrappers to `QueryBuilder`. All tests pass, but I don't know if what I am doing is a good design. To me the "bug" behavior was the `QueryBuilder` was returning an instance of itself for chaining even if the underlying knex call returned a value and ended chaining.

thetutlage commented:
Here you go https://github.com/benallfree/adonis-lucid/pull/1

benallfree commented:
I addressed #226 too. Please review #226 notes and carefully review this commit, because I changed how aggregate queries are prepared. All tests pass, I think this is the way to go https://github.com/adonisjs/adonis-lucid/pull/205/commits/302b03eff7a5b2293e989bc729441bdaf1cfe990

thetutlage commented:
@benallfree Sorry it took a while to look into it. I 100% reviewed the PR, but there are some merge conflicts now, you can please fix them and I will merge it.

Also all code from `develop` inside merge conflicts is required, since that was added after your PR

benallfree commented:
Conflicts fixed, please give it one final check! :)

thetutlage commented:
Thanks @benallfree for keeping the patience
